### PR TITLE
Alerting: new min_interval_seconds options to enforce a minimum eval frequency

### LINKED ISSUE #14447: Feature request: Alert, minimum time
### What Grafana version are you using?
5.3.4
### What datasource are you using?
Influx
### What OS are you running grafana on?
Debian

Possibility to set in admin interface the minimum time for an alert to check(Evaulate every).

Ttoday a user with rights to create alerts, can with no knowledge of the system, set an alert to check every 1s, which mean unnecessary many checks.

It would have been nice if an admin can set that the value cannot be less than xx seconds/minute etc.

### PR DESCRIPTION
**Pull request by uepoch**
Fixes: #14447

It will allow operator to force alerting scheduler to honor a minimum frequency when Tick() is called

### CODE REVIEW

File: pkg/setting/setting.go [hunk 1]
```diff
@@ -895,6 +896,7 @@ func (cfg *Cfg) Load(args *CommandLineArgs) error {
 	notificationTimeoutSeconds := alerting.Key("notification_timeout_seconds").MustInt64(30)
 	AlertingNotificationTimeout = time.Second * time.Duration(notificationTimeoutSeconds)
 	AlertingMaxAttempts = alerting.Key("max_attempts").MustInt(3)
+	AlertingMinInterval = alerting.Key("min_interval_seconds").MustInt64(0)
```
Comment (bergquist) [inline pkg/setting/setting.go]:
```suggestion
	AlertingMinInterval = alerting.Key("min_interval_seconds").MustInt64(100)
```

**Review by jfwm2 — APPROVED** [review]

**Review by jfwm2 — APPROVED** [review]

**Review by jfwm2 — APPROVED** [review]

**Review by bergquist — COMMENTED** [review]
Thank you for taking the time to contribute to Grafana!

Looks good but it could be slightly confusing for the users since the UI never indicates that the interval cannot be below the min value. Adding some input validations in https://github.com/grafana/grafana/blob/master/pkg/services/alerting/extractor.go#L113 would be a good idea.

**Comment by uepoch** [comment]
While I 100% agree about the UX problem, this was more to be able to throttle grafana when metric system is overloaded (ponctual event), not preventing people to add specific values, so when we lower the limit, they don't have to change the value

I would rather go for a warning sign at Alert creation/display time, to inform users about the on-going limitation

My frontend skills are not the best, but how hard would it be to add a message with proper warn feeling in the Alert tab, saying "Configuration of your grafana server will enforce at least X seconds between evals" ?

**Comment by bergquist** [comment]
Showing more descriptive text int the GUI would be nice but I think it would be enough to just add some validation in `extractor.go` that I linked to. If it returns an error it will be shown in Grafana when someone tries to save a dashboard.

**Comment by uepoch** [comment]
Wouldn't it force all users to change their alerting frequency value before being able to save something that is not related to alerting ?
I think this feature can be used as both a baseline for the grafana cluster (so, in that case having the validator is fine), but also as a reactive strategy for operators when you need to limit the number of calls done by grafana. For the latter, we would like it to not disturb regular user workflows since it's a temporary change

Maybe we can extend the validator idea with another flag in the config ? if passed, alerts can't be created/updated if the value is under the threshold, if not it's an invisible throttling in grafana backend

WDYT ?

**Comment by bergquist** [comment]
My first idea was to use the same setting for both purposes. But that would be weird if you open a dashboard with alerts, change something not related to alerts, save it and get rejected cause the frequency is not high enough. 

On the other side. Two settings feel a bit too much to me. It would only make sense during the transition. 

Perhaps the best way is to only implement the validation. My guess is that this case is only a problem for large installations where operators should be able to write a SQL update statement that updates frequency on all alerts rules with too low frequency. 

Any thoughts?

**Comment by uepoch** [comment]
Sorry for the long waiting, I had to context switch :' ( 

I kinda agree with you that 2 flags is too much, it feels clunky.

At the same time, we don't want to force a validation (for now), as it's an irreversible change in our clients habits and we want to be able to limit the grafana server itself, more than our users, as it creates a constant load (and a huge one..) that we would like to reduce during unstable periods.

I understand your point in the long run about having a "minimal freq" verification made in the front-end, but I don't think it's solving the same issue.
Currently there's no way to reduce Grafana's alerting impact instantly without relying on either cutting all alerts, or doing some dirty modifications inside the DB (for which you have a whole new world of issues, with indempotence and such) which doesn't really feel great either

While I think validation based on config is a great nice to have, it feels like it could be included in a more thoughtful management system where more precise restrictions could be applied (per org, per user, .... ) for not only alerting but also auto-refresh, annotations/variables/etc

Would that latter system make sense to you @bergquist ?

**Comment by bergquist** [comment]
I've been very back and forth about this feature but I think it's fine to introduce this if we give the users some information about it on the alert tab. Otherwise, its gonna be confusing for users who expect alert rules to be evaluated every second but the backend restricts it to every hour. 

The information in the alert tab could be shown if `min_interval_seconds` is not set to 1 (which should be the default).

File: conf/defaults.ini [hunk 2]
```diff
@@ -506,6 +506,9 @@ notification_timeout_seconds = 30
 # Default setting for max attempts to sending alert notifications. Default value is 3
 max_attempts = 3
 
+# Makes it possible to enforce a minimal interval between evaluations, to reduce load on the backend
+min_interval_seconds = 60
```
Comment (bergquist) [inline conf/defaults.ini]: the default value should `1`

**Comment by uepoch** [comment]
Sorry for the extra delay, I've modified it as your required

**Comment by marefr** [comment]
@uepoch sorry for the delay on reviewing this. Have you considered fixing this?

> give the users some information about it on the alert tab. Otherwise, its gonna be confusing for users who expect alert rules to be evaluated every second but the backend restricts it to every hour.
> 
> The information in the alert tab could be shown if min_interval_seconds is not set to 1 (which should be the default).

Let me know if you need any help with that

**Comment by uepoch** [comment]
Sorry it went out of my mind, I will take a look

My frontend skills are reaaaaally not up to the challenge tho, if you have hints about where to look I would gladly help with it

**Comment by marefr** [comment]
@uepoch sorry it taken long time for me to come back to this. 

Here you can add your setting among some of the other alerting settings and then it will be exposed to the browser:
https://github.com/grafana/grafana/blob/00d0640b6e778ddaca021670fe851fe00982acf2/pkg/api/frontendsettings.go#L176-L178

Similarly, you'll need change here to be able to read the config client side:
https://github.com/grafana/grafana/blob/00d0640b6e778ddaca021670fe851fe00982acf2/packages/grafana-runtime/src/config.ts#L30-L32

Here's the angular/html part where the frequency input is defined:
https://github.com/grafana/grafana/blob/00d0640b6e778ddaca021670fe851fe00982acf2/public/app/features/alerting/partials/alert_tab.html#L16

Here you can set the default frequency and also an example of how `config` values are read:
https://github.com/grafana/grafana/blob/00d0640b6e778ddaca021670fe851fe00982acf2/public/app/features/alerting/AlertTabCtrl.ts#L187-L189

Let me know if you'll need more input and/or help.

**Comment by papagian** [comment]
@uepoch We would like to go forward with this and merge it.
Are you interested in making the remaining modifications [proposed](https://github.com/grafana/grafana/pull/17279#issuecomment-542178232) by @marefr.

**Comment by papagian** [comment]
Closing this is favour of #21188 which is the continuation of the work done here.
Unfortunately I was unable to push commits to this one.

**Pull request closed without merging**